src/ripple/app/consensus/RCLConsensus.h

```diff
@@ -71,318 +359,98 @@ class RCLConsensus final : public Consensus<RCLConsensus, RCLCxTraits>,
-        @param ledger The ledger at the time of the state change
-        @param haveCorrectLCL Whether we believ we have the correct LCL.
-    */
-    void
-    notify(
-        protocol::NodeEvent ne,
-        RCLCxLedger const& ledger,
-        bool haveCorrectLCL);
-
-    /** Accept a new ledger based on the given transactions.
-
-        @ref onAccept
-     */
-    void
-    doAccept(
-        Result const& result,
-        RCLCxLedger const& prevLedger,
-        NetClock::duration closeResolution,
-        CloseTimes const& rawCloseTimes,
-        Mode const& mode);
-
-    /** Build the new last closed ledger.
-
-        Accept the given the provided set of consensus transactions and build
-        the last closed ledger. Since consensus just agrees on which
-        transactions to apply, but not whether they make it into the closed
-        ledger, this function also populates retriableTxs with those that can
-        be retried in the next round.
-
-        @param previousLedger Prior ledger building upon
-        @param set The set of transactions to apply to the ledger
-        @param closeTime The the ledger closed
-        @param closeTimeCorrect Whether consensus agreed on close time
-        @param closeResolution Resolution used to determine consensus close time
-        @param roundTime Duration of this consensus rorund
-        @param retriableTxs Populate with transactions to retry in next round
-        @return The newly built ledger
-  */
-    RCLCxLedger
-    buildLCL(
-        RCLCxLedger const& previousLedger,
-        RCLTxSet const& set,
-        NetClock::time_point closeTime,
-        bool closeTimeCorrect,
-        NetClock::duration closeResolution,
-        std::chrono::milliseconds roundTime,
-        CanonicalTXSet& retriableTxs);
-
-    /** Validate the given ledger and share with peers as necessary
-
-        @param ledger The ledger to validate
-        @param proposing Whether we were proposing transactions while generating
-                         this ledger.  If we are not proposing, a validation
-                         can still be sent to inform peers that we know we
-                         aren't fully participating in consensus but are still
-                         around and trying to catch up.
-    */
-    void
-    validate(RCLCxLedger const& ledger, bool proposing);
-
-    //!-------------------------------------------------------------------------
-    Application& app_;
-    std::unique_ptr<FeeVote> feeVote_;
-    LedgerMaster& ledgerMaster_;
-    LocalTxs& localTxs_;
-    InboundTransactions& inboundTransactions_;
+    peerProposal(
+        NetClock::time_point const& now,
+        RCLCxPeerPos const& newProposal);
+private:
+    // Since Consensus does not provide intrinsic thread-safety, this mutex
+    // guards all calls to consensus_. adaptor_ uses atomics internally
+    // to allow concurrent access of its data members that have getters.
+    mutable std::recursive_mutex mutex_;
```

Comment:
It is generally best to avoid `recursive_mutex` if possible.  However I don't loose sleep over the use of `recursive_mutex` (I put it in the standard ;-) ).  I can whip up a shrieking mutex if we decide that's what we want. :-)